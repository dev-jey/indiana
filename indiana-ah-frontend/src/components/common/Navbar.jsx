import React, { Fragment, Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { Navbar, Nav } from 'react-bootstrap';
import SignupContainer from '../SignupFormContainer.jsx';
import LoginContainer from '../LoginFormContainer.jsx';
import ResetContainer from '../ResetFormContainer.jsx';
import Modal from './Modal.jsx';
import { ProfileImg, ImageLogo } from '../../styles/styledComponents/Navigation.jsx';
import { customLogo } from '../../assets/images/svg';
import Button from '../../styles/styledComponents/Button.jsx';
import { signOutUser } from '../../redux/actions/authActions';
import Dropdown from '../Dropdown.jsx';
import SearchBar from './SearchBar.jsx';

export class NavBar extends Component {
  state = {
    dropDown: false,
    modalIsOpen: false,
    modalContent: ''
  };

  setUsername = (userData) => {
    const { name, username } = userData;
    let displayName;
    if (name) {
      const display = name.split(' ')[0];
      displayName = display;
    } else if (username) {
      displayName = username.replace(/\d{5,}/, '');
    } else {
      displayName = ' ';
    }
    return displayName;
  };

  openModal = () => {
    this.setState(() => ({ modalIsOpen: true }));
  };

  closeModal = () => {
    const {
      auth: { isLoading }
    } = this.props;
    if (!isLoading) this.setState(() => ({ modalIsOpen: false }));
  };

  displayForm = (form) => {
    this.setState({ modalIsOpen: true, modalContent: form });
  };

  componentWillUnmount() {
    document.removeEventListener('click', this.closeDropDown);
  }

  dropDown = (event) => {
    event.preventDefault();

    this.setState({ dropDown: true }, () => {
      document.addEventListener('click', this.closeDropDown);
    });
  };

  closeDropDown = () => {
    this.setState({ dropDown: false }, () => {
      document.removeEventListener('click', this.closeDropDown);
    });
  };

  render() {
    const { user, auth } = this.props;
    const { modalContent } = this.state;
    const userLInk = (
      <Nav className="d-flex flex-row justify-content-between">
        <ProfileImg
          src={user.userData.profileImage || user.userData.imageUrl}
          className="ml-5 mt-2"
          alt="logo"
        />
        <div to="/signup" className="d-flex ft-size-2 ml-5">
          <span className="username ml-3">
            {this.setUsername(user.userData)}
          </span>
          <span style={{ padding: '0 0.3em' }} />
          <Dropdown signOutUser={this.props.signOutUser} />
        </div>
      </Nav>
    );
    const form = () => {
      switch (modalContent) {
        case 'login':
          return <LoginContainer
            displayForm={this.displayForm}
            closeModal={this.closeModal}
          />;
        case 'register':
          return <SignupContainer
            displayForm={this.displayForm}
            closeModal={this.closeModal}
          />;
        case 'reset':
          return <ResetContainer
            displayForm={this.displayForm}
            closeModal={this.closeModal}
          />;
        default:
          return null;
      }
    };
    return (
      <Fragment>
        <Navbar bg="white shadow-sm px-5" expand="lg">
          <Navbar.Brand>
            <Link to="/">
              <ImageLogo>
                <img src={customLogo} className="feather" />
              </ImageLogo>
            </Link>
          </Navbar.Brand>
          <Navbar.Toggle aria-controls="basic-navbar-nav" />
          <Navbar.Collapse id="basic-navbar-nav">
            <Nav className="mr-auto ml-5">
              <SearchBar />
            </Nav>
            {auth.isAuthenticated ? (
              userLInk
            ) : (
              <Button onClick={() => this.displayForm('login')} className="ml-auto" sm>
                Login
              </Button>
            )}
          </Navbar.Collapse>
        </Navbar>
        <Modal
          modalIsOpen={this.state.modalIsOpen}
          closeModal={this.closeModal}
          body={form()}
        />
      </Fragment>
    );
  }
}

NavBar.propTypes = {
  user: PropTypes.object,
  signOutUser: PropTypes.func,
  auth: PropTypes.object
};

export const mapStateToProps = state => ({
  user: state.user,
  auth: state.auth
});

export default connect(
  mapStateToProps,
  { signOutUser }
)(NavBar);
