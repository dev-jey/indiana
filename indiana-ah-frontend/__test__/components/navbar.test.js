import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import SignupContainer from '../../src/components/SignupFormContainer.jsx';
import LoginContainer from '../../src/components/LoginFormContainer.jsx';
import ResetContainer from '../../src/components/ResetFormContainer.jsx';
import Navbar, { NavBar } from '../../src/components/common/Navbar.jsx';

const initialState = {
  auth: { isAuthenticated: false },
  user: { userData: { username: 'ozone4real' } }
};

const mockStore = configureStore([thunk]);
const store = mockStore(initialState);
const mockFn = jest.fn();
const event = {
  preventDefault: jest.fn()
};

describe('Custom Navbar component test', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter>
            <Navbar />
          </MemoryRouter>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders a navbar', () => {
    const navbar = mount(
      <Provider store={store}>
        <MemoryRouter>
          <Navbar />
        </MemoryRouter>
      </Provider>
    );
    expect(navbar.find('.navbar')).toBeDefined();
    const inputField = navbar.find('input');
    const button = navbar.find('button');
    const button2 = navbar.find('button').at(1);
    button2.simulate('click');
    expect(inputField.length).toEqual(1);
    expect(button.length).toEqual(2);
  });

  it('tests for its functionalities', () => {
    const wrapper = shallow(<NavBar {...initialState} signOutUser={jest.fn()} />);
    wrapper.instance().openModal();
    expect(wrapper.state().modalIsOpen).toEqual(true);
    wrapper.instance().closeModal(mockFn);
    expect(wrapper.state().modalIsOpen).toEqual(false);
    wrapper.instance().displayForm('login');
    expect(wrapper.find(LoginContainer)).toBeDefined();
    wrapper.instance().displayForm('register');
    expect(wrapper.find(SignupContainer)).toBeDefined();
    wrapper.instance().displayForm('reset');
    expect(wrapper.find(ResetContainer)).toBeDefined();
    wrapper.instance().dropDown(event);
    wrapper.instance().closeDropDown();
    wrapper.instance().componentWillUnmount();
    wrapper.setProps({ auth: { isAuthenticated: true } });
    wrapper.setState({ dropDown: true });
  });
});
