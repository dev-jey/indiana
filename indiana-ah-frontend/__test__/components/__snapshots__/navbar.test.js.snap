// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Navbar component test renders correctly 1`] = `
Array [
  <nav
    className="navbar navbar-expand-lg navbar-light bg-white shadow-sm px-5"
  >
    <span
      className="navbar-brand"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <div
          className="sc-bZQynM hVJJCu"
        >
          <img
            className="feather"
            src="test-file-stub"
          />
        </div>
      </a>
    </span>
    <button
      aria-controls="basic-navbar-nav"
      aria-label="Toggle navigation"
      className="navbar-toggler"
      onClick={[Function]}
      type="button"
    >
      <span
        className="navbar-toggler-icon"
      />
    </button>
    <div
      aria-expanded={null}
      className="navbar-collapse collapse"
      id="basic-navbar-nav"
    >
      <div
        className="mr-auto ml-5 navbar-nav"
        onKeyDown={[Function]}
      >
        <div
          className="search-bar"
        >
          <form
            className=""
            onSubmit={[Function]}
          >
            <div
              className="mb-3 input-group"
            >
              <select
                name="filterOption"
                onChange={[Function]}
              >
                <option
                  value=""
                >
                  All
                </option>
                <option
                  value="author"
                >
                  Author
                </option>
                <option
                  value="tag"
                >
                  Tag
                </option>
                <option
                  value="title"
                >
                  Title
                </option>
              </select>
              <input
                aria-describedby="basic-addon2"
                className="form-control"
                name="searchValue"
                onChange={[Function]}
                placeholder="Search Articles"
                required={true}
                value=""
              />
              <div
                className="input-group-append"
              >
                <span
                  className="input-group-text"
                  id="basic-addon2"
                  onClick={[Function]}
                >
                  <i
                    className="fa fa-search"
                  />
                </span>
              </div>
            </div>
          </form>
        </div>
      </div>
      <button
        className="ml-auto sc-htpNat dIJQYP"
        onClick={[Function]}
      >
        Login
      </button>
    </div>
  </nav>,
  <div />,
]
`;
